use scatter_add for index_add when dim is the most inner dim

CaoE opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #88729

### Motivation
When dim is -1 and the slice of source or result is noncontiguous, original `index_add` is slow as it uses add for the sliced tensor, which is serial on index and parallel on sliced tensor to avoid write conflict. Doing parallel on the sliced tensor is not optimal as the size of sliced tensor may be not big enough to parallel and also causes multiple parallelizations.

`scatter_add ` is used to speedup for this case as `scatter_add ` parallels on the outer dimension of input and is serial on the inner dimension to avoid write conflict. `scatter_add ` only need one parallel and the size of outer dimensions is bigger to do parallel.


### Testing

- Single core:

Before:

shape | fp32 / s | bf16 / s
-- | -- | --
[10, 128, 20, 20] | 2.82E-03 | 2.11E-03
[10, 128, 50, 50] | 0.023604 | 0.023794


After:

shape | fp32 / s | bf16 / s
-- | -- | --
[10, 128, 20, 20] | 9.30E-04 | 1.66E-03
[10, 128, 50, 50] | 0.005995 | 0.010003

- Single socket (28 cores):

Before:

shape | fp32 / s | bf16 / s
-- | -- | --
[10, 128, 20, 20] | 2.96E-03 | 2.52E-03
[10, 128, 50, 50] | 0.012208 | 0.012568

After:

shape | fp32 / s | bf16 / s
-- | -- | --
[10, 128, 20, 20] | 7.44E-05 | 1.33E-04
[10, 128, 50, 50] | 0.000333 | 0.000469






cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10

CaoE committed:
use scatter_add for index_add when dim is the most inner dim

[ghstack-poisoned]

CaoE committed:
Update on "use scatter_add for index_add when dim is the most inner dim"


### Motivation
When dim is -1 and the slice of source or result is noncontiguous, original `index_add` is slow as it uses add for the sliced tensor which will use scalar add and can not be paralleled due to write conflict. Therefore, `scatter_add ` is used to speedup for this case as the optimized `scatter_add ` can handle the write conflict and fully parallel it.

[ghstack-poisoned]

CaoE committed:
Update on "use scatter_add for index_add when dim is the most inner dim"


### Motivation
When dim is -1 and the slice of source or result is noncontiguous, original `index_add` is slow as it uses add for the sliced tensor which will use scalar add and can not be paralleled due to write conflict. Therefore, `scatter_add ` is used to speedup for this case as the optimized `scatter_add ` can handle the write conflict and fully parallel it.


[ghstack-poisoned]

CaoE committed:
Update on "use scatter_add for index_add when dim is the most inner dim"


### Motivation
When dim is -1 and the slice of source or result is noncontiguous, original `index_add` is slow as it uses add for the sliced tensor which will use scalar add and can not be paralleled due to write conflict. Therefore, `scatter_add ` is used to speedup for this case as the optimized `scatter_add ` can handle the write conflict and fully parallel it.


[ghstack-poisoned]

mingfeima commented on test/test_torch.py:
```diff
@@ -5671,6 +5671,40 @@ def test_index_add_all_dtypes(self):
+        # dim is -1, alpha is 1, and dtype of index is torch.long,
+        # i.e., using scatter_add
+        for dtype in all_types_and_complex_and(torch.half, torch.bfloat16):
+            if torch.cuda.is_available():
```
add `devices` for the test case and try to compare on conditions such as `alpha = 1.0` and `alpha = 2.0 and result /= 2.0`.

mingfeima commented:
Need more benchmark data for this one.

CaoE committed:
Update on "use scatter_add for index_add when dim is the most inner dim"


### Motivation
When dim is -1 and the slice of source or result is noncontiguous, original `index_add` is slow as it uses add for the sliced tensor which will use scalar add and can not be fully paralleled due to write conflict. Therefore, `scatter_add ` is used to speedup for this case as the optimized `scatter_add ` can handle the write conflict and fully parallel it.


cc VitalyFedyunin jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

CaoE committed:
Update on "use scatter_add for index_add when dim is the most inner dim"


### Motivation
When dim is -1 and the slice of source or result is noncontiguous, original `index_add` is slow as it uses add for the sliced tensor which will use scalar add and can not be fully paralleled due to write conflict. Therefore, `scatter_add ` is used to speedup for this case as the optimized `scatter_add ` can handle the write conflict and fully parallel it.



[ghstack-poisoned]

CaoE committed:
Update on "use scatter_add for index_add when dim is the most inner dim"


### Motivation
When dim is -1 and the slice of source or result is noncontiguous, original `index_add` is slow as it uses add for the sliced tensor which will use scalar add and can not be fully paralleled due to write conflict. Therefore, `scatter_add ` is used to speedup for this case as the optimized `scatter_add ` can handle the write conflict and fully parallel it.



[ghstack-poisoned]

CaoE committed:
Update on "use scatter_add for index_add when dim is the most inner dim"


### Motivation
When dim is -1 and the slice of source or result is noncontiguous, original `index_add` is slow as it uses add for the sliced tensor which will use scalar add and can not be fully paralleled due to write conflict. Therefore, `scatter_add ` is used to speedup for this case as the optimized `scatter_add ` can handle the write conflict and fully parallel it.



[ghstack-poisoned]

CaoE committed:
Update on "use scatter_add for index_add when dim is the most inner dim"


### Motivation
When dim is -1 and the slice of source or result is noncontiguous, original `index_add` is slow as it uses add for the sliced tensor, which is serial on the outer dimension and parallel on the most inner dimension to avoid write conflict. Doing parallel on the inner dimension is not optimal as the size of inner dimension is not big enough to parallel and also causes multiple parallelizations.

`scatter_add ` is used to speedup for this case as `scatter_add ` parallels on the outer dimension and is serial on the inner dimension to avoid write conflict. `scatter_add ` only need one parallel and the size of outer dimensions is bigger to do parallel.


### Testing

- Single core:

Before:

shape | fp32 / s | bf16 / s
-- | -- | --
[10, 128, 20, 20] | 2.82E-03 | 2.11E-03
[10, 128, 50, 50] | 0.023604 | 0.023794


After:

shape | fp32 / s | bf16 / s
-- | -- | --
[10, 128, 20, 20] | 9.30E-04 | 1.66E-03
[10, 128, 50, 50] | 0.005995 | 0.010003

- Single socket (28 cores):

Before:

shape | fp32 / s | bf16 / s
-- | -- | --
[10, 128, 20, 20] | 2.96E-03 | 2.52E-03
[10, 128, 50, 50] | 0.012208 | 0.012568

After:

shape | fp32 / s | bf16 / s
-- | -- | --
[10, 128, 20, 20] | 7.44E-05 | 1.33E-04
[10, 128, 50, 50] | 0.000333 | 0.000469






[ghstack-poisoned]

jgong5 commented on aten/src/ATen/native/TensorAdvancedIndexing.cpp:
```diff
@@ -858,6 +858,33 @@ TORCH_IMPL_FUNC(index_add_cpu_out)
     if (numel == 0) {
       return;
     }
+    if (dim == (result.dim() - 1) && alpha.equal(1.0) &&
```
Does `dim` has to be the last dim here or this fast path can also run faster with other dim values?

CaoE replied:
When https://github.com/pytorch/pytorch/pull/82703 lands, using scatter_add will also get obvious speedup for the case dim == 0. Will add the cases dim == 0 after https://github.com/pytorch/pytorch/pull/82703 lands.

jgong5 commented on aten/src/ATen/native/TensorAdvancedIndexing.cpp:
```diff
@@ -858,6 +858,33 @@ TORCH_IMPL_FUNC(index_add_cpu_out)
     }
+    if (dim == (result.dim() - 1) && alpha.equal(1.0) &&
+        index_contig.scalar_type() == ScalarType::Long &&
+        // Heuristic value that scatter_add has higher performance.
+        result.numel() >= (4 * at::internal::GRAIN_SIZE / at::get_num_threads())) {
```
Could you please explain the logic of this heuristics?

CaoE replied:
Use `result.numel() > at::internal::GRAIN_SIZE` instread to make sure `scatter_add`  has enough data to get relatively good performance.

CaoE committed:
Update on "use scatter_add for index_add when dim is the most inner dim"


### Motivation
When dim is -1 and the slice of source or result is noncontiguous, original `index_add` is slow as it uses add for the sliced tensor, which is serial on index and parallel on sliced tensor to avoid write conflict. Doing parallel on the sliced tensor is not optimal as the size of sliced tensor may be not big enough to parallel and also causes multiple parallelizations.

`scatter_add ` is used to speedup for this case as `scatter_add ` parallels on the outer dimension of input and is serial on the inner dimension to avoid write conflict. `scatter_add ` only need one parallel and the size of outer dimensions is bigger to do parallel.


### Testing

- Single core:

Before:

shape | fp32 / s | bf16 / s
-- | -- | --
[10, 128, 20, 20] | 2.82E-03 | 2.11E-03
[10, 128, 50, 50] | 0.023604 | 0.023794


After:

shape | fp32 / s | bf16 / s
-- | -- | --
[10, 128, 20, 20] | 9.30E-04 | 1.66E-03
[10, 128, 50, 50] | 0.005995 | 0.010003

- Single socket (28 cores):

Before:

shape | fp32 / s | bf16 / s
-- | -- | --
[10, 128, 20, 20] | 2.96E-03 | 2.52E-03
[10, 128, 50, 50] | 0.012208 | 0.012568

After:

shape | fp32 / s | bf16 / s
-- | -- | --
[10, 128, 20, 20] | 7.44E-05 | 1.33E-04
[10, 128, 50, 50] | 0.000333 | 0.000469






[ghstack-poisoned]

CaoE committed:
Update on "use scatter_add for index_add when dim is the most inner dim"


### Motivation
When dim is -1 and the slice of source or result is noncontiguous, original `index_add` is slow as it uses add for the sliced tensor, which is serial on index and parallel on sliced tensor to avoid write conflict. Doing parallel on the sliced tensor is not optimal as the size of sliced tensor may be not big enough to parallel and also causes multiple parallelizations.

`scatter_add ` is used to speedup for this case as `scatter_add ` parallels on the outer dimension of input and is serial on the inner dimension to avoid write conflict. `scatter_add ` only need one parallel and the size of outer dimensions is bigger to do parallel.


### Testing

- Single core:

Before:

shape | fp32 / s | bf16 / s
-- | -- | --
[10, 128, 20, 20] | 2.82E-03 | 2.11E-03
[10, 128, 50, 50] | 0.023604 | 0.023794


After:

shape | fp32 / s | bf16 / s
-- | -- | --
[10, 128, 20, 20] | 9.30E-04 | 1.66E-03
[10, 128, 50, 50] | 0.005995 | 0.010003

- Single socket (28 cores):

Before:

shape | fp32 / s | bf16 / s
-- | -- | --
[10, 128, 20, 20] | 2.96E-03 | 2.52E-03
[10, 128, 50, 50] | 0.012208 | 0.012568

After:

shape | fp32 / s | bf16 / s
-- | -- | --
[10, 128, 20, 20] | 7.44E-05 | 1.33E-04
[10, 128, 50, 50] | 0.000333 | 0.000469






[ghstack-poisoned]

CaoE commented on aten/src/ATen/native/TensorAdvancedIndexing.cpp:
```diff
@@ -858,6 +858,33 @@ TORCH_IMPL_FUNC(index_add_cpu_out)
+      // Check whether result and source are matched apart from the dimension dim.
+      // Note that the broadcast case:
+      // source.select(dim, i) is broadcast for result.select(dim, index_data[i])
+      // The broadcast case is not applicable for scatter_add
```
add test cases for the broadcast case.

mingfeima commented:
I suppose we need to update the parallel reduce algorithm (for dim reduction) in Tensoriterator eventually, this is currently a temp solution.

CaoE committed:
Update on "use scatter_add for index_add when dim is the most inner dim"


### Motivation
When dim is -1 and the slice of source or result is noncontiguous, original `index_add` is slow as it uses add for the sliced tensor, which is serial on index and parallel on sliced tensor to avoid write conflict. Doing parallel on the sliced tensor is not optimal as the size of sliced tensor may be not big enough to parallel and also causes multiple parallelizations.

`scatter_add ` is used to speedup for this case as `scatter_add ` parallels on the outer dimension of input and is serial on the inner dimension to avoid write conflict. `scatter_add ` only need one parallel and the size of outer dimensions is bigger to do parallel.


### Testing

- Single core:

Before:

shape | fp32 / s | bf16 / s
-- | -- | --
[10, 128, 20, 20] | 2.82E-03 | 2.11E-03
[10, 128, 50, 50] | 0.023604 | 0.023794


After:

shape | fp32 / s | bf16 / s
-- | -- | --
[10, 128, 20, 20] | 9.30E-04 | 1.66E-03
[10, 128, 50, 50] | 0.005995 | 0.010003

- Single socket (28 cores):

Before:

shape | fp32 / s | bf16 / s
-- | -- | --
[10, 128, 20, 20] | 2.96E-03 | 2.52E-03
[10, 128, 50, 50] | 0.012208 | 0.012568

After:

shape | fp32 / s | bf16 / s
-- | -- | --
[10, 128, 20, 20] | 7.44E-05 | 1.33E-04
[10, 128, 50, 50] | 0.000333 | 0.000469






[ghstack-poisoned]

CaoE committed:
Update on "use scatter_add for index_add when dim is the most inner dim"


### Motivation
When dim is -1 and the slice of source or result is noncontiguous, original `index_add` is slow as it uses add for the sliced tensor, which is serial on index and parallel on sliced tensor to avoid write conflict. Doing parallel on the sliced tensor is not optimal as the size of sliced tensor may be not big enough to parallel and also causes multiple parallelizations.

`scatter_add ` is used to speedup for this case as `scatter_add ` parallels on the outer dimension of input and is serial on the inner dimension to avoid write conflict. `scatter_add ` only need one parallel and the size of outer dimensions is bigger to do parallel.


### Testing

- Single core:

Before:

shape | fp32 / s | bf16 / s
-- | -- | --
[10, 128, 20, 20] | 2.82E-03 | 2.11E-03
[10, 128, 50, 50] | 0.023604 | 0.023794


After:

shape | fp32 / s | bf16 / s
-- | -- | --
[10, 128, 20, 20] | 9.30E-04 | 1.66E-03
[10, 128, 50, 50] | 0.005995 | 0.010003

- Single socket (28 cores):

Before:

shape | fp32 / s | bf16 / s
-- | -- | --
[10, 128, 20, 20] | 2.96E-03 | 2.52E-03
[10, 128, 50, 50] | 0.012208 | 0.012568

After:

shape | fp32 / s | bf16 / s
-- | -- | --
[10, 128, 20, 20] | 7.44E-05 | 1.33E-04
[10, 128, 50, 50] | 0.000333 | 0.000469






[ghstack-poisoned]

CaoE committed:
Update on "use scatter_add for index_add when dim is the most inner dim"


### Motivation
When dim is -1 and the slice of source or result is noncontiguous, original `index_add` is slow as it uses add for the sliced tensor, which is serial on index and parallel on sliced tensor to avoid write conflict. Doing parallel on the sliced tensor is not optimal as the size of sliced tensor may be not big enough to parallel and also causes multiple parallelizations.

`scatter_add ` is used to speedup for this case as `scatter_add ` parallels on the outer dimension of input and is serial on the inner dimension to avoid write conflict. `scatter_add ` only need one parallel and the size of outer dimensions is bigger to do parallel.


### Testing

- Single core:

Before:

shape | fp32 / s | bf16 / s
-- | -- | --
[10, 128, 20, 20] | 2.82E-03 | 2.11E-03
[10, 128, 50, 50] | 0.023604 | 0.023794


After:

shape | fp32 / s | bf16 / s
-- | -- | --
[10, 128, 20, 20] | 9.30E-04 | 1.66E-03
[10, 128, 50, 50] | 0.005995 | 0.010003

- Single socket (28 cores):

Before:

shape | fp32 / s | bf16 / s
-- | -- | --
[10, 128, 20, 20] | 2.96E-03 | 2.52E-03
[10, 128, 50, 50] | 0.012208 | 0.012568

After:

shape | fp32 / s | bf16 / s
-- | -- | --
[10, 128, 20, 20] | 7.44E-05 | 1.33E-04
[10, 128, 50, 50] | 0.000333 | 0.000469






[ghstack-poisoned]

CaoE commented:
> I suppose we need to update the parallel reduce algorithm (for dim reduction) in Tensoriterator eventually, this is currently a temp solution.

Yes, this solution has the same  potential precision issues for bfloat16 as the original implementation since multiple values may be accumulated to same location in bfloat16. We can merge this PR first and then submit another PR to solve such flaw for pytorch.

CaoE committed:
Update on "use scatter_add for index_add when dim is the most inner dim"


### Motivation
When dim is -1 and the slice of source or result is noncontiguous, original `index_add` is slow as it uses add for the sliced tensor, which is serial on index and parallel on sliced tensor to avoid write conflict. Doing parallel on the sliced tensor is not optimal as the size of sliced tensor may be not big enough to parallel and also causes multiple parallelizations.

`scatter_add ` is used to speedup for this case as `scatter_add ` parallels on the outer dimension of input and is serial on the inner dimension to avoid write conflict. `scatter_add ` only need one parallel and the size of outer dimensions is bigger to do parallel.


### Testing

- Single core:

Before:

shape | fp32 / s | bf16 / s
-- | -- | --
[10, 128, 20, 20] | 2.82E-03 | 2.11E-03
[10, 128, 50, 50] | 0.023604 | 0.023794


After:

shape | fp32 / s | bf16 / s
-- | -- | --
[10, 128, 20, 20] | 9.30E-04 | 1.66E-03
[10, 128, 50, 50] | 0.005995 | 0.010003

- Single socket (28 cores):

Before:

shape | fp32 / s | bf16 / s
-- | -- | --
[10, 128, 20, 20] | 2.96E-03 | 2.52E-03
[10, 128, 50, 50] | 0.012208 | 0.012568

After:

shape | fp32 / s | bf16 / s
-- | -- | --
[10, 128, 20, 20] | 7.44E-05 | 1.33E-04
[10, 128, 50, 50] | 0.000333 | 0.000469






cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

CaoE commented:
@pytorchbot merge

CaoE commented:
@kit1980 Could you please review this PR ? Thank you.

CaoE commented:
@malfet Could you please review this PR ? Thank you.

malfet commented on aten/src/ATen/native/TensorAdvancedIndexing.cpp:
```diff
@@ -859,6 +859,58 @@ TORCH_IMPL_FUNC(index_add_cpu_out)
+        auto ep_index = index_contig.as_strided(ep_sizes, ep_strides);
+        result.scatter_add_(dim, ep_index, source);
+        return;
+      };
```
Nit
```suggestion
      }
```

CaoE replied:
Fixed it.

malfet commented on aten/src/ATen/native/TensorAdvancedIndexing.cpp:
```diff
@@ -859,6 +859,58 @@ TORCH_IMPL_FUNC(index_add_cpu_out)
+        int64_t ndim = a.size();
+        for (int64_t i = ndim - 1; i >= 0; --i) {
+          if (i != dim){
+            if (a[i] != b[i])
+              return false;
```
```suggestion
            if (a[i] != b[i]) {
              return false;
           }   
```

CaoE replied:
Fixed it.

malfet commented on aten/src/ATen/native/TensorAdvancedIndexing.cpp:
```diff
@@ -859,6 +859,58 @@ TORCH_IMPL_FUNC(index_add_cpu_out)
+      // source.select(dim, i) is broadcast for result.select(dim, index_data[i])
+      // The broadcast case is not applicable for scatter_add
+      auto check_sizes = [&ep_sizes, &ep_strides](IntArrayRef a, IntArrayRef b, int64_t dim) -> bool {
+        if (a.size() != b.size())
+          return false;
```
```suggestion
        if (a.size() != b.size()) {
          return false;
        }  
```

CaoE replied:
Fixed it.

malfet commented on aten/src/ATen/native/TensorAdvancedIndexing.cpp:
```diff
@@ -859,6 +859,58 @@ TORCH_IMPL_FUNC(index_add_cpu_out)
+          return false;
+        int64_t ndim = a.size();
+        for (int64_t i = ndim - 1; i >= 0; --i) {
+          if (i != dim){
```
Nit. (whitespace after curly bracket) Also, please consider dismantling pyramid of doom by replacing this with `if (i == dim) { continue; }`
```suggestion
          if (i != dim) {
```

CaoE replied:
Fixed it.

malfet commented on aten/src/ATen/native/TensorAdvancedIndexing.cpp:
```diff
@@ -859,6 +859,58 @@ TORCH_IMPL_FUNC(index_add_cpu_out)
+        if (a.size() != b.size())
+          return false;
+        int64_t ndim = a.size();
+        for (int64_t i = ndim - 1; i >= 0; --i) {
```
Why this loop must be from `ndim-1` to 0 rather than conventional `0` to `ndim-1`? (This way one can also get rid of `ndim` variable)
```suggestion
        for (const auto i : c10::irange(a.size())) {
```

CaoE replied:
Fixed it.

malfet commented on aten/src/ATen/native/TensorAdvancedIndexing.cpp:
```diff
@@ -859,6 +859,58 @@ TORCH_IMPL_FUNC(index_add_cpu_out)
+        result.scalar_type() != ScalarType::ComplexHalf) {
+      std::vector<int64_t> ep_sizes(result.sizes().size());
+      std::vector<int64_t> ep_strides(source.sizes().size());
+      ep_sizes[dim] = numel;
+      ep_strides[dim] = 1;
```
Why not move this to the loop body inside of lambda? This way entire ep_size/ep_strides initialization logic would be located in the same place

CaoE replied:
Fixed it.

malfet commented on aten/src/ATen/native/TensorAdvancedIndexing.cpp:
```diff
@@ -859,6 +859,61 @@ TORCH_IMPL_FUNC(index_add_cpu_out)
+        ep_sizes[dim] = numel;
+        ep_strides[dim] = 1;
+        for (const auto i : c10::irange(a.size())) {
+          if (i == dim) { continue; }
```
```suggestion
          if (i == dim) { 
              continue;
          }
```

CaoE replied:
Fixed it.

CaoE commented:
@pytorchbot merge

desertfire commented:
@pytorchbot revert -m "causing inductor test failure" -c nosignal

desertfire commented:
Please use the `ciflow/inductor` label to test relanding.

CaoE committed:
Update on "use scatter_add for index_add when dim is the most inner dim"


### Motivation
When dim is -1 and the slice of source or result is noncontiguous, original `index_add` is slow as it uses add for the sliced tensor, which is serial on index and parallel on sliced tensor to avoid write conflict. Doing parallel on the sliced tensor is not optimal as the size of sliced tensor may be not big enough to parallel and also causes multiple parallelizations.

`scatter_add ` is used to speedup for this case as `scatter_add ` parallels on the outer dimension of input and is serial on the inner dimension to avoid write conflict. `scatter_add ` only need one parallel and the size of outer dimensions is bigger to do parallel.


### Testing

- Single core:

Before:

shape | fp32 / s | bf16 / s
-- | -- | --
[10, 128, 20, 20] | 2.82E-03 | 2.11E-03
[10, 128, 50, 50] | 0.023604 | 0.023794


After:

shape | fp32 / s | bf16 / s
-- | -- | --
[10, 128, 20, 20] | 9.30E-04 | 1.66E-03
[10, 128, 50, 50] | 0.005995 | 0.010003

- Single socket (28 cores):

Before:

shape | fp32 / s | bf16 / s
-- | -- | --
[10, 128, 20, 20] | 2.96E-03 | 2.52E-03
[10, 128, 50, 50] | 0.012208 | 0.012568

After:

shape | fp32 / s | bf16 / s
-- | -- | --
[10, 128, 20, 20] | 7.44E-05 | 1.33E-04
[10, 128, 50, 50] | 0.000333 | 0.000469






cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

CaoE committed:
Update on "use scatter_add for index_add when dim is the most inner dim"


### Motivation
When dim is -1 and the slice of source or result is noncontiguous, original `index_add` is slow as it uses add for the sliced tensor, which is serial on index and parallel on sliced tensor to avoid write conflict. Doing parallel on the sliced tensor is not optimal as the size of sliced tensor may be not big enough to parallel and also causes multiple parallelizations.

`scatter_add ` is used to speedup for this case as `scatter_add ` parallels on the outer dimension of input and is serial on the inner dimension to avoid write conflict. `scatter_add ` only need one parallel and the size of outer dimensions is bigger to do parallel.


### Testing

- Single core:

Before:

shape | fp32 / s | bf16 / s
-- | -- | --
[10, 128, 20, 20] | 2.82E-03 | 2.11E-03
[10, 128, 50, 50] | 0.023604 | 0.023794


After:

shape | fp32 / s | bf16 / s
-- | -- | --
[10, 128, 20, 20] | 9.30E-04 | 1.66E-03
[10, 128, 50, 50] | 0.005995 | 0.010003

- Single socket (28 cores):

Before:

shape | fp32 / s | bf16 / s
-- | -- | --
[10, 128, 20, 20] | 2.96E-03 | 2.52E-03
[10, 128, 50, 50] | 0.012208 | 0.012568

After:

shape | fp32 / s | bf16 / s
-- | -- | --
[10, 128, 20, 20] | 7.44E-05 | 1.33E-04
[10, 128, 50, 50] | 0.000333 | 0.000469






cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

CaoE committed:
Update on "use scatter_add for index_add when dim is the most inner dim"


### Motivation
When dim is -1 and the slice of source or result is noncontiguous, original `index_add` is slow as it uses add for the sliced tensor, which is serial on index and parallel on sliced tensor to avoid write conflict. Doing parallel on the sliced tensor is not optimal as the size of sliced tensor may be not big enough to parallel and also causes multiple parallelizations.

`scatter_add ` is used to speedup for this case as `scatter_add ` parallels on the outer dimension of input and is serial on the inner dimension to avoid write conflict. `scatter_add ` only need one parallel and the size of outer dimensions is bigger to do parallel.


### Testing

- Single core:

Before:

shape | fp32 / s | bf16 / s
-- | -- | --
[10, 128, 20, 20] | 2.82E-03 | 2.11E-03
[10, 128, 50, 50] | 0.023604 | 0.023794


After:

shape | fp32 / s | bf16 / s
-- | -- | --
[10, 128, 20, 20] | 9.30E-04 | 1.66E-03
[10, 128, 50, 50] | 0.005995 | 0.010003

- Single socket (28 cores):

Before:

shape | fp32 / s | bf16 / s
-- | -- | --
[10, 128, 20, 20] | 2.96E-03 | 2.52E-03
[10, 128, 50, 50] | 0.012208 | 0.012568

After:

shape | fp32 / s | bf16 / s
-- | -- | --
[10, 128, 20, 20] | 7.44E-05 | 1.33E-04
[10, 128, 50, 50] | 0.000333 | 0.000469






cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

CaoE commented:
Hi @desertfire, ciflow/inductor label is added and all checks have passed.  Will it cause inductor test failure again when the PR is landed ?

CaoE commented:
@pytorchbot merge -r

atalman commented:
@pytorchbot revert -m "Break internal build" -c ghfirst

atalman commented:
Following error is observer on internal build:
```
fbcode/caffe2/aten/src/ATen/native/TensorAdvancedIndexing.cpp:897:17: error: comparison of integers of different signs: 'const unsigned long' and 'int64_t' (aka 'long') [-Werror,-Wsign-compare]
          if (i == dim) {
              ~ ^  ~~~
```

Please try compile the code with ```-Wsign-compare```. Please resolve this issue and resubmit this PR. 
Casting i to int64_t or creating new int64_t resolves this issue:
```
if (int64_t(i) == dim)
```

CaoE committed:
Update on "use scatter_add for index_add when dim is the most inner dim"


### Motivation
When dim is -1 and the slice of source or result is noncontiguous, original `index_add` is slow as it uses add for the sliced tensor, which is serial on index and parallel on sliced tensor to avoid write conflict. Doing parallel on the sliced tensor is not optimal as the size of sliced tensor may be not big enough to parallel and also causes multiple parallelizations.

`scatter_add ` is used to speedup for this case as `scatter_add ` parallels on the outer dimension of input and is serial on the inner dimension to avoid write conflict. `scatter_add ` only need one parallel and the size of outer dimensions is bigger to do parallel.


### Testing

- Single core:

Before:

shape | fp32 / s | bf16 / s
-- | -- | --
[10, 128, 20, 20] | 2.82E-03 | 2.11E-03
[10, 128, 50, 50] | 0.023604 | 0.023794


After:

shape | fp32 / s | bf16 / s
-- | -- | --
[10, 128, 20, 20] | 9.30E-04 | 1.66E-03
[10, 128, 50, 50] | 0.005995 | 0.010003

- Single socket (28 cores):

Before:

shape | fp32 / s | bf16 / s
-- | -- | --
[10, 128, 20, 20] | 2.96E-03 | 2.52E-03
[10, 128, 50, 50] | 0.012208 | 0.012568

After:

shape | fp32 / s | bf16 / s
-- | -- | --
[10, 128, 20, 20] | 7.44E-05 | 1.33E-04
[10, 128, 50, 50] | 0.000333 | 0.000469






cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

CaoE committed:
Update on "use scatter_add for index_add when dim is the most inner dim"


### Motivation
When dim is -1 and the slice of source or result is noncontiguous, original `index_add` is slow as it uses add for the sliced tensor, which is serial on index and parallel on sliced tensor to avoid write conflict. Doing parallel on the sliced tensor is not optimal as the size of sliced tensor may be not big enough to parallel and also causes multiple parallelizations.

`scatter_add ` is used to speedup for this case as `scatter_add ` parallels on the outer dimension of input and is serial on the inner dimension to avoid write conflict. `scatter_add ` only need one parallel and the size of outer dimensions is bigger to do parallel.


### Testing

- Single core:

Before:

shape | fp32 / s | bf16 / s
-- | -- | --
[10, 128, 20, 20] | 2.82E-03 | 2.11E-03
[10, 128, 50, 50] | 0.023604 | 0.023794


After:

shape | fp32 / s | bf16 / s
-- | -- | --
[10, 128, 20, 20] | 9.30E-04 | 1.66E-03
[10, 128, 50, 50] | 0.005995 | 0.010003

- Single socket (28 cores):

Before:

shape | fp32 / s | bf16 / s
-- | -- | --
[10, 128, 20, 20] | 2.96E-03 | 2.52E-03
[10, 128, 50, 50] | 0.012208 | 0.012568

After:

shape | fp32 / s | bf16 / s
-- | -- | --
[10, 128, 20, 20] | 7.44E-05 | 1.33E-04
[10, 128, 50, 50] | 0.000333 | 0.000469






cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

CaoE commented:
@pytorchbot merge

CaoE committed:
Update on "use scatter_add for index_add when dim is the most inner dim"


### Motivation
When dim is -1 and the slice of source or result is noncontiguous, original `index_add` is slow as it uses add for the sliced tensor, which is serial on index and parallel on sliced tensor to avoid write conflict. Doing parallel on the sliced tensor is not optimal as the size of sliced tensor may be not big enough to parallel and also causes multiple parallelizations.

`scatter_add ` is used to speedup for this case as `scatter_add ` parallels on the outer dimension of input and is serial on the inner dimension to avoid write conflict. `scatter_add ` only need one parallel and the size of outer dimensions is bigger to do parallel.


### Testing

- Single core:

Before:

shape | fp32 / s | bf16 / s
-- | -- | --
[10, 128, 20, 20] | 2.82E-03 | 2.11E-03
[10, 128, 50, 50] | 0.023604 | 0.023794


After:

shape | fp32 / s | bf16 / s
-- | -- | --
[10, 128, 20, 20] | 9.30E-04 | 1.66E-03
[10, 128, 50, 50] | 0.005995 | 0.010003

- Single socket (28 cores):

Before:

shape | fp32 / s | bf16 / s
-- | -- | --
[10, 128, 20, 20] | 2.96E-03 | 2.52E-03
[10, 128, 50, 50] | 0.012208 | 0.012568

After:

shape | fp32 / s | bf16 / s
-- | -- | --
[10, 128, 20, 20] | 7.44E-05 | 1.33E-04
[10, 128, 50, 50] | 0.000333 | 0.000469






cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

CaoE commented:
@pytorchbot merge

CaoE committed:
Update on "use scatter_add for index_add when dim is the most inner dim"


### Motivation
When dim is -1 and the slice of source or result is noncontiguous, original `index_add` is slow as it uses add for the sliced tensor, which is serial on index and parallel on sliced tensor to avoid write conflict. Doing parallel on the sliced tensor is not optimal as the size of sliced tensor may be not big enough to parallel and also causes multiple parallelizations.

`scatter_add ` is used to speedup for this case as `scatter_add ` parallels on the outer dimension of input and is serial on the inner dimension to avoid write conflict. `scatter_add ` only need one parallel and the size of outer dimensions is bigger to do parallel.


### Testing

- Single core:

Before:

shape | fp32 / s | bf16 / s
-- | -- | --
[10, 128, 20, 20] | 2.82E-03 | 2.11E-03
[10, 128, 50, 50] | 0.023604 | 0.023794


After:

shape | fp32 / s | bf16 / s
-- | -- | --
[10, 128, 20, 20] | 9.30E-04 | 1.66E-03
[10, 128, 50, 50] | 0.005995 | 0.010003

- Single socket (28 cores):

Before:

shape | fp32 / s | bf16 / s
-- | -- | --
[10, 128, 20, 20] | 2.96E-03 | 2.52E-03
[10, 128, 50, 50] | 0.012208 | 0.012568

After:

shape | fp32 / s | bf16 / s
-- | -- | --
[10, 128, 20, 20] | 7.44E-05 | 1.33E-04
[10, 128, 50, 50] | 0.000333 | 0.000469






cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]
